cirq/value/linear_dict.py

```diff
@@ -0,0 +1,173 @@
+
+    For convenience, methods which take a second linear combination accept
+    a standalone key instance as well. The second argument is then treated as
+    a single-term linear combination with the sole coefficient equal to 1.0.
+    """
+    def __init__(self, terms: Mapping[Any, Scalar]) -> None:
+        """Initializes linear combination from a collection of terms.
+
+        Args:
+            terms: Mapping of abstract vectors to coefficients in the linear
+                combination being initialized.
+        """
+        super(LinearDict, self).__init__()
+        self.update(terms)
+
+    def clean(self, *, tolerance: float=1e-9) -> None:
+        """Modifies self by deleting terms with small coefficients."""
+        negligible = [v for v, c in self.items() if abs(c) <= tolerance]
+        for v in negligible:
+            del self[v]
+
+    def copy(self) -> 'LinearDict':
+        return LinearDict(super(LinearDict, self).copy())
+
+    def __getitem__(self, vector: Any) -> Scalar:
+        if vector in self:
+            return super(LinearDict, self).__getitem__(vector)
+        return 0.0
+
+    def __iadd__(self, other: Any) -> 'LinearDict':
+        if not isinstance(other, LinearDict):
+            other = LinearDict({other: 1})
+
+        for vector, coefficient in other.items():
+            self[vector] = self[vector] + coefficient
+        self.clean(tolerance=0)
+        return self
+
+    def __add__(self, other: Any) -> 'LinearDict':
+        result = self.copy()
+        result += other
+        return result
+
+    def __isub__(self, other: Any) -> 'LinearDict':
+        if not isinstance(other, LinearDict):
+            other = LinearDict({other: 1})
+
+        for vector, coefficient in other.items():
+            self[vector] = self[vector] - coefficient
+        self.clean(tolerance=0)
+        return self
+
+    def __sub__(self, other: Any) -> 'LinearDict':
+        result = self.copy()
+        result -= other
+        return result
+
+    def __neg__(self) -> 'LinearDict':
+        return LinearDict({v: -c for v, c in self.items()})
+
+    def __imul__(self, a: Scalar) -> 'LinearDict':
+        for vector in self:
+            self[vector] *= a
+        self.clean(tolerance=0)
+        return self
+
+    def __mul__(self, a: Scalar) -> 'LinearDict':
+        result = self.copy()
+        result *= a
+        return result
+
+    def __rmul__(self, a: Scalar) -> 'LinearDict':
+        return self.__mul__(a)
+
+    def __truediv__(self, a: Scalar) -> 'LinearDict':
+        return self.__mul__(1 / a)
+
+    def __bool__(self) -> bool:
+        return not all(c == 0 for c in self.values())
+
+    def __eq__(self, other: Any) -> bool:
+        """Checks whether two linear combinations are exactly equal.
+
+        Presence or absence of terms with coefficients exactly equal to
+        zero does not affect outcome.
+
+        Not appropriate for most practical purposes due to sensitivity to
```

Comment:
I would quickly scan through all of python's magic methods to see if we missed any more.

Reply:
Yeah, there is more, e.g. `__len__()` etc. Will check carefully.

Reply:
```
In [1]: dir(dict())
Out[1]: 
['__class__',                       # override not needed
 '__contains__',                    # done
 '__delattr__',                     # override not needed
 '__delitem__'                      # override not needed
 '__dir__',                         # override not needed
 '__doc__',                         # set from docstring
 '__eq__',                          # done
 '__format__',                      # done
 '__ge__',                          # override not needed
 '__getattribute__',                # override not needed
 '__getitem__',                     # done
 '__gt__',                          # override not needed
 '__hash__',                        # override not needed
 '__init__',                        # done
 '__init_subclass__',               # override not needed
 '__iter__',                        # done
 '__le__',                          # override not needed
 '__len__',                         # done
 '__lt__',                          # override not needed
 '__ne__',                          # done
 '__new__',                         # override not needed
 '__reduce__',                      # override not needed
 '__reduce_ex__',                   # override not needed
 '__repr__',                        # done
 '__setattr__',                     # override not needed
 '__setitem__',                     # done
 '__sizeof__',                      # override not needed
 '__str__',                         # done
 '__subclasshook__',                # override not needed
 'clear',                           # override not needed
 'copy',                            # done
 'fromkeys',                        # done
 'get',                             # done
 'items',                           # done
 'keys',                            # done
 'pop',                             # override not needed
 'popitem',                         # override not needed
 'setdefault',                      # override not needed
 'update',                          # done
 'values']                          # done
```